### DISCUSSION TITLE
Package Manager improvements to show installed and versions

### DISCUSSION THREAD

**Discussion by deefrawley** [Flow-Launcher/Flow.Launcher#523] [Ideas]
Package Manager suggestions:

1. A new command from within Flow to show what plugins are installed and their versions. Something like `pm list'
2. Include package version numbers in the `pm install` search results listing

**Comment by jjw24** [comment]
> A new command from within Flow to show what plugins are installed and their versions. Something like `pm list'

we have a pm uninstall, maybe we rename that to pm list, show version, and uninstall when you select the plugin. what do you think?

**Comment by deefrawley** [comment]
I think it is unintuitive to have the select function uninstall as the user might just be looking for more info about the plugin until they learn of the context menu. Speaking of which, perhaps make uninstall one of the options there?

**Comment by taooceros** [comment]
Maybe make uninstall lies within the context menu?

**Comment by jjw24** [comment]
> Maybe make uninstall lies within the context menu?

Sounds good, what do you think @deefrawley ?

**Comment by deefrawley** [comment]
Yep, good spot for it.
Taking this further there is the possibility to just have one `pm` command. It lists all plugins, with installed ones at the top of the list (with some sort of indicator like "installed" text or icon or both). Installed with update waiting can also be shown. Not installed plugins are below these. Context menu is used to install/uninstall depending on the current state.
The only potential problem here is that as far as I can tell the context menu is right click mouse only. No keyboard shortcut. Which means the mouse is required to manage plugins. I think we would need to have ways to open the context menu with the keyboard.

**Comment by deefrawley** [comment]
> The only potential problem here is that as far as I can tell the context menu is right click mouse only.

As I've learned on the Discord, not it most certainly isn't! My bad.
What about the suggestion of the one command package manager where all actions are completed via the context menu?

**Comment by jjw24** [comment]
> What about the suggestion of the one command package manager where all actions are completed via the context menu?

Sorry, what do you mean?

**Comment by pc223** [comment]
> What about the suggestion of the one command package manager where all actions are completed via the context menu?

@deefrawley Like only use `pm` without postfix `install`, `uninstall`,...?

Steps:
1. `pm` to list all plugins
2. `Ctrl+N`, `Ctrl+P` to select plugin
3. `Shift+Enter` to open Context Menu
4. `Ctrl+N/P` to choose `install, uninstall, ...`

Is this what you mean? 😃 

Seems interesting, but I'm not too sure about the UX of this, but it's worth to try, I think we can "beta test" this new mode for a while to see if it's good UX

**Comment by taooceros** [comment]
@pc223 If you would like to try to finish this.
Here's some hint.
1. No need for UI related (xaml) change, but modify returned results.
2. Check out Plugins.PluginsManager.Main class (which implemented IAsyncPlugin interface) and the QueryAsync method. The return value will be the things displayed when typing `pm`.
3. ContextMenu related things lies in class ContextMenu.cs, which has implemented IContextMenu interface.

**Comment by pc223** [comment]
> @pc223 If you would like to try to finish this.
> Here's some hint.
> 
> 1. No need for UI related (xaml) change, but modify returned results.
> 2. Check out Plugins.PluginsManager.Main class (which implemented IAsyncPlugin interface) and the QueryAsync method. The return value will be the things displayed when typing `pm`.
> 3. ContextMenu related things lies in class ContextMenu.cs, which has implemented IContextMenu interface.

Thanks for the hints, it'll be useful when someone tries this

But I'm not planning to do this any time soon haha 😄 I don't know much C# and not familiar with how `pm` designed, so doing this will cost me at least around x3 time as C#-dev

Right now I kind of focus on something I can fix relatively quick, low risk, small fish stuff haha, maybe try big fish in the future 😼

**Comment by deefrawley** [comment]
No worries. Maybe one for v1.9? Having thought about it I think it should be bundled under `pm list` just to future proof as there may be another `pm` argument needed in the future.
I still think plugin version numbers should be added sooner rather than later. Hoping that's not too hard to fit into the current implementation.
